oauth2/src/main/java/com/linecorp/armeria/common/auth/oauth2/AbstractOAuth2Request.java

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.auth.oauth2;
+
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.client.WebClient;
+import com.linecorp.armeria.common.AggregatedHttpResponse;
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.common.QueryParams;
+import com.linecorp.armeria.common.QueryParamsBuilder;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.RequestHeadersBuilder;
+
+/**
+ * A common abstraction for the requests implementing various OAuth 2.0 request/response flows,
+ * as per <a href="https://tools.ietf.org/html/rfc6749">[RFC6749]</a> and other relevant specifications.
+ * @param <T> the type of the authorization result.
+ */
+public abstract class AbstractOAuth2Request<T> {
+
+    private final WebClient endpoint;
+    private final String endpointPath;
+    @Nullable
+    private final ClientAuthorization clientAuthorization;
+
+    /**
+     * A common abstraction for the requests implementing various Authorization request/response flows,
+     * as per <a href="https://tools.ietf.org/html/rfc6749">[RFC6749]</a> and other relevant specifications.
+     *
+     * @param endpoint A {@link WebClient} to facilitate the Authorization requests. Must
+     *                 correspond to the required Authorization endpoint of the OAuth 2 system.
+     * @param endpointPath A URI path that corresponds to the Authorization endpoint of the
+     *                     OAuth 2 system.
+     * @param clientAuthorization Provides client authorization for the OAuth requests,
+     *                            as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>.
+     */
+    protected AbstractOAuth2Request(WebClient endpoint, String endpointPath,
+                                    @Nullable ClientAuthorization clientAuthorization) {
```

Comment:
I see both `*Request` and `*Grant` require `WebClient` as an endpoint. How about letting a user specify a `WebClient` when creating an `OAuth2Client` or `OAuth2*Authorizer` and changing `*Request` and `*Grant` to require a `WebClient` when sending a request? This way, we can keep only one `WebClient` in `OAuth2Client` or `OAuth2*Authorizer`.

Or, is there a use case for specifying different `WebClient` for each request?

Reply:
Requests aren't public. Grant passes WebClient to Requests whenever appropriate. So there are no multiple WebClients.
>> is there a use case for specifying different WebClient for each request
It all depends on the particular Grant. The Grant decides (as per the spec) whether to use the same endpoint/WebClient for different operations/requests or not.

Reply:
OK!